### PR TITLE
[adsp] cleanup system start way

### PR DESCRIPTION
**Pull request by AlwinEsch**
Have currently no idea which other Application Messages instead of "on" and "off" required for ADSP, that's why I removed the use of `public KODI::MESSAGING::IMessageTarget` on CActiveAEDSP and passed everything to CApplication to reduce the needed code.

If OK for you and no other ideas for them in DSP itself then I merge it in.

### CODE REVIEW

**Comment by FernetMenta** [comment]
@Paxxi please review

**Comment by Razzeee** [comment]
Think he said he has a lot of pressure on work, at the moment and is unlikely to show up here in the next time. So he might not show up :/

**Comment by da-anda** [comment]
isn't this change going a bit into the wrong direction? I thought we strive for decoupling things while this is actually the opposite (it was cleanly decoupled and self contained from what I see and now adds IMO unrelated logic to cApplication). It's ofc less overall code and only a couple lines are added to cApp that don't do any harm, but at least to me it feels wrong. This is by no means an objection to this PR and it's up to the senior/architecture devs to decide anyways, I only wanted to share my thoughts.

**Comment by AlwinEsch** [comment]
Can understand you and you also right, thought only by the use of enable and disable only is it a bit small to separate and seen PVRManager also this way. Thought is OK ;-)

**Comment by AlwinEsch** [comment]
Maybe I can go the other way and leave it there, currently I cleanup also the PVRManager startup and make this also the way it was on DSP.

My Goal is to make the binary add-on systems to the same style to handle them on end with same base class, related to web binary add-on system which brings the third bigger code system.

**Comment by FernetMenta** [comment]
g_AEDSPManager needs to go away. globals are evil.

**Comment by AlwinEsch** [comment]
Sorry @FernetMenta, how you mean this? Mean you to use the current change style on this request?

**Comment by AlwinEsch** [comment]
Here are my current changes to PVR https://github.com/AlwinEsch/kodi/commit/13f42b3008d2adc04175257f4ac198943901c7fe (not finished)

**Comment by FernetMenta** [comment]
The primary purpose of application messenger was to have something executed by the main thread. The current design is dead lock happy and should IMO not be expanded. CApplicationMessenger::GetInstance().SendMsg blocks the calling thread. If this thread is needed to finish the task, it will block forever. Globals make this even worse because those can be called from everywhere and the caller does not know what threads are involved.
Bottom line: avoid app messenger and globals

**Comment by Paxxi** [comment]
I vote no on this, we don't want to move stuff back into capplication unless absolutely necessary.
@FernetMenta as for your concerns, it may have started out as an off thread task runner but I think with current design and maybe some expansion we can use it as a generalized messaging system to break up some of our entangled code. I'm not saying it's the answer to everything and it should be considered among the other options.

Deadlocks are a concern but we do try to minimize the risk by holding no internal locks in messenger during calls so I don't think the risk is higher than other cross thread calling methods. Whenever possible one should try to use PostMsg as it's always asynchronous and returns immediately.

**Comment by FernetMenta** [comment]
@Paxxi this has nothing to do with holding locks. A one-way synchronous messaging system is not the way to go for the reasons I gave and in particular not in direction of Application. CApplication has dependencies to almost everything, hence it is very likely that CApplication waits on something after a message. The problem is that the calling thread is blocked.
Either messaging is in both directions with no blocking or into a components that does not make any calls outside.

Example:
Call SendMsg(TMSG_MEDIA_STOP) from dvdplayer thread and the app is dead. There are masses of those cases.

**Comment by AlwinEsch** [comment]
After my PVR system changes are done (https://github.com/AlwinEsch/kodi/commits/pvr-cleanup) and fixed the crash there if PVR system becomes disabled with TV channel on playback.

I'm try later to remove the work complete from CApplication to prevent problems on Main Thread.

**Comment by Paxxi** [comment]
@FernetMenta that's why I mentioned using PostMsg whenever possible.

Does this code actually need to use messaging or would regular method calls work just as well? (Not in front of a computer)

Also I'm not against the idea of two way message passing/reply but there's a lot of work even to adapt actor stuff we have into something usable as most of our code relies on synchronous method calls

**Comment by FernetMenta** [comment]
@Paxxi 

> that's why I mentioned using PostMsg whenever possible.

I agree but what are the non possible cases and how can we get rid of those?

**Comment by AlwinEsch** [comment]
Just a note the new system comes now in on PVR. Related to the comments above PVR has exactly this problem with messages.

If PVR becomes disabled during playback calls CApplication ```CPVRManager::GetInstance().Stop();``` and this want to stop playback on CApplication which not possible until he return and brings a crash.

**Comment by AchimTuran** [comment]
@AlwinEsch @FernetMenta @Paxxi @da-anda 
At the first time I saw this PR, I thought I did something wrong during the merge of AudioDSP.

As discussed on DevCon it would be nice if we can create a place to share visions and ideas for Kodi. It would also be nice to share the ideas why was something changed to a new structure. It's now the second time this week that I saw where someone is working on old stuff.

**Comment by AlwinEsch** [comment]
Closed this, related changes becomes included to another one. Temporary commit here visible: https://github.com/AlwinEsch/kodi/commit/78854e214ac6b87d0406b329ade7d4249c66e543
